redis/commands/del.cc

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C) 2019 pengjian.uestc @ gmail.com
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "redis/commands/del.hh"
+#include "seastar/core/shared_ptr.hh"
+#include "redis/request.hh"
+#include "redis/reply.hh"
+#include "service_permit.hh"
+#include "types.hh"
+#include "service/storage_proxy.hh"
+#include "service/client_state.hh"
+#include "redis/options.hh"
+#include "redis/mutation_utils.hh"
+
+namespace redis {
+
+namespace commands {
+
+shared_ptr<abstract_command> del::prepare(service::storage_proxy& proxy, request&& req)
+{
+    if (req.arguments_size() != 1) {
+        throw wrong_arguments_exception(1, req.arguments_size(), req._command);
+    }
+    return seastar::make_shared<del> (std::move(req._command), std::move(req._args[0]));
+}
+
+future<redis_message> del::execute(service::storage_proxy& proxy, redis::redis_options& options, service_permit permit)
+{
+    return redis::delete_object(proxy, options, std::move(_key), permit).then([] {
+       return redis_message::one();
```

Comment:
Currently your implement only supports to delete one key once.
Redis supports to delete multiple keys once: https://redis.io/commands/del

It returns the number of deleted keys, not always to be one: `":1\r\n"`

Reply:
zero `":0\r\n"` should be returned if the key doesn't exist.

Reply:
It's easy to delete multi keys. 
In fact, we should determine that how many keys should be actually deleted. This is a read-before-write operation. As we known that Scylla does not support LWT. So should we just read all keys, then delete the exist keys?

Reply:
> It's easy to delete multi keys.
> In fact, we should determine that how many keys should be actually deleted. This is a read-before-write operation. As we known that Scylla does not support LWT. So should we just read all keys, then delete the exist keys?

@psarna , @nyh can you help to answer @fastio's question?

@fastio , Scylla had supported lwt: https://github.com/scylladb/scylla/issues/1359

Reply:
This is an interesting question. Indeed Scylla's classic MO is to do writes **without read**, so you can delete a key without knowing whether it actually existed before or it did not. Redis's DEL documentation also says that it "Removes the specified keys. A key is ignored if it does not exist.", but it is indeed sad that the API also returns the number of keys actually removed. I don't know if users actually use this capability often, but sadly it is officially documented and we'll eventually need to support it. 

But we don't need to support it **now**. What I suggest we do now is to remember that we still have this problem - I suggest doing two things: First have an "xfailing" test for this case (a test deleting a non-existent key expecting the returned integer to be 0) which will succeed on real redis, but fail on Scylla. Second is to create a bug-tracker issue to remember this.

For multiple keys, again, if you don't want to fix this issue immediately, please add an xfailing test and an issue to remember this missing feature. Or if it's trivial, just fix that now.

Eventually, we'll need to decide how to do all sorts of RMW (read-modify-write aka read-before-write) operations that Redis has. Maybe we'll use LWT, maybe some sort of leader model or Raft, or I don't know what; Scylla's DynamoDB support ("Alternator") has the same problems. But we can start without this for now. We just need to remember what we don't support yet, and in my opinion xfailing tests (which succeed on the real redis) is a good way to remember that.